### PR TITLE
Added configuration for case_sensitive option in StringFilter

### LINKED ISSUE #334: proposal: case insensitive search on postgres
for now, we do this:
https://gist.github.com/dbu/9524776#file-caseinsensitivestringfilter-php-L40

doctrine_orm_filter_types.xml add an option to configure the default
behaviour for sonata.admin.orm.filter.type.string
additionally, we could add an option to the filter to set it per case
(who knows, maybe i have one field where i really want case sensitive
filtering).

btw, one can achieve the opposite, make mysql case sensitive for likes
with  "column COLLATE utf8_bin LIKE :value" (resp = or NOT LIKE)

not sure how this is best modeled. if we don't care about BC, i would
say postgres should by default start to compare case insensitive. but
this would be very surprising for existing applications that rely on the
behaviour.

having 2 more ChoiceType options sounds like a bad idea too, it will
overkill for the admin user most of the time.

interested in a PR to add an option to the DI configuration to explicitly set case sensitivity (with 3 states: default, case-sensitive, case-insensitive), plus a per-filter option.

i don't know about other db than postgres + mysql, so i would throw an exception that this is not supported if an explicit option is set, so that people at least know something is not working.

### PR DESCRIPTION
**Pull request by Nadrshin**
## Subject

I am targeting this branch, because I need an opportunity to set a case-insensitive search for the StringFilter through configuration.

Closes #334

## Changelog

```markdown
### Added
- Added configuration for `case_sensitive` option in `StringFilter`
```

### CODE REVIEW

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -76,6 +83,20 @@ public function getConfigTreeBuilder()
                         ->end()
                     ->end()
                 ->end()
+                ->arrayNode('filters')
```
Comment (core23) [inline src/DependencyInjection/Configuration.php]: Can you please move this to a new method

File: src/Filter/StringFilter.php [hunk 2]
```diff
@@ -24,6 +24,19 @@ class StringFilter extends Filter
...
 
+    /**
+     * @var bool
+     */
+    private $caseSensitive;
+
+    /**
+     * @param bool $caseSensitive
+     */
+    public function __construct($caseSensitive)
```
Comment (core23) [inline src/Filter/StringFilter.php]:
```suggestion
    public function __construct(bool $caseSensitive = false)
```

File: src/DependencyInjection/Configuration.php (hunk 1 above)
Comment (Nadrshin) [reply src/DependencyInjection/Configuration.php]: I added method `getFiltersNode`.

File: src/Filter/StringFilter.php (hunk 2 above)
Comment (Nadrshin) [reply src/Filter/StringFilter.php]: Done. But I used `true` as default value.

**Comment by Nadrshin** [comment]
@core23, I provided unit tests.

It is my first PR in github. Therefore, there may be problems.

PS: Maybe I should rename the variable `StringFilter::$caseSensitive` to `StringFilter::$defaultCaseSensitive`?

**Comment by core23** [comment]
> @core23, I provided unit tests.
> 
> It is my first PR in github. 

Welcome to the family :)

> 
> PS: Maybe I should rename the variable `StringFilter::$caseSensitive` to `StringFilter::$defaultCaseSensitive`?

Good catch! This would be a better name.

File: src/Filter/StringFilter.php [hunk 3]
```diff
@@ -24,6 +24,16 @@ class StringFilter extends Filter
         ChoiceType::TYPE_EQUAL => '=',
     ];
 
+    /**
+     * @var bool
+     */
+    private $defaultCaseSensitive;
+
+    public function __construct(bool $defaultCaseSensitive = true)
```
Comment (OskarStark) [inline src/Filter/StringFilter.php]: If someone sets it it’s not default anymore.
I would go without the default prefix

Comment (core23) [reply src/Filter/StringFilter.php]: It's just to keep BC for users who extend this class

**Comment by VincentLanglet** [comment]
The tests fail weirdly. You should try to commit amend and force push to re-trigger the run.

File: src/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -77,8 +78,39 @@ public function getConfigTreeBuilder()
...
+            ->append($this->getFiltersNode())
         ;
 
         return $treeBuilder;
     }
+
+    private function getFiltersNode(): ArrayNodeDefinition
+    {
+        $treeBuilder = new TreeBuilder();
+        $node = $treeBuilder->root('filters');
```
Comment (franmomu) [inline src/DependencyInjection/Configuration.php]:
```suggestion
        $treeBuilder = new TreeBuilder('filters');
        $node = $treeBuilder->getRootNode();
```

**Comment by greg0ire** [comment]
@VincentLanglet it depends on the RDBMS. On Postgres, case sensitivity is the default IIRC.

**Comment by VincentLanglet** [comment]
> @VincentLanglet it depends on the RDBMS. On Postgres, case sensitivity is the default IIRC.

IMO, It should not rely on this. It is weird having a `case sensitive` option if it's overriden by the database.

Shouldn't we add something like
```
if ($this->getOption('case_sensitive')) {
    $or->add(sprintf('BINARY %s.%s %s :%s', $alias, $field, $operator, $parameterName));
} else {
    $or->add(sprintf('BINARY LOWER(%s.%s) %s :%s', $alias, $field, $operator, $parameterName));
}
```

**Comment by greg0ire** [comment]
I was going to argue that functions make indices unusable and should be avoided, but using `LIKE` with '%' at the beginning of a pattern do that too… is it what's happening in your case? What happens when you search for `EST`? Does `Testa` come up? If yes, we're probably not using indices already.

**Comment by VincentLanglet** [comment]
> I was going to argue that functions make indices unusable and should be avoided, but using `LIKE` with '%' at the beginning of a pattern do that too… is it what's happening in your case? What happens when you search for `EST`? Does `Testa` come up? If yes, we're probably not using indices already.

image

**Comment by greg0ire** [comment]
Ok well we might as well use functions everywhere then, since performance is already ruined :sweat_smile:

**Comment by VincentLanglet** [comment]
The fix would be better with this feature (in this PR or in another), because I consider this like a bug, but it can be considered like a breaking chance for sql database users.
Giving an option to override easily the behaviour could be great.

@Nadrshin Do you have time to finish your PR ? :)

**Comment by Nadrshin** [comment]
Sorry, I don't have time.

**Pull request closed without merging**